assignments/defsym needs to be processed in link order

quic-seaswara opened the pull request:
It seems like assignments and defsym expressions need to be processed by link order, if the expression refers to a symbol that is not yet visited by the linker, the symbol will need to be processed as an undefined symbol. This means that any expression that refers to a symbol that can be PROVIDED by the linker script can fill in the symbol or resolve to that symbol.

This particular combination is exercised when building picolibc.

This bug allows picolibc to successfully build/run tests when tested on RISC-V.

Closes #60

Change-Id: I98d94d78cf415f28aacf04df0da10983642d9b41

partaror commented on lib/Script/Assignment.cpp:
```diff
@@ -264,3 +261,9 @@ bool Assignment::isDot() const { return (Name.size() == 1 && Name[0] == '.'); }
   return isDot() || ExpressionToEvaluate->hasDot();
 }
+
+std::unordered_set<std::string> Assignment::getSymbolNames() const {
```
We have a very similar existing function 'Assignment::getSymbolNames(std::unordered_set<std::string> &SymbolTokens)`. Can we use that function to implement this new function? This way we will not duplicate how we are computing the symbol names. This being said, it is only one line in both the functions -- `ExpressionToEvaluate->getSymbolNames(SymbolTokens);`

partaror commented on include/eld/SymbolResolver/NamePool.h:
```diff
@@ -129,6 +129,11 @@ class NamePool {
 
+  // Helper function to add an undefined symbol to the NamePool
+  void
+  addUndefinedELFSymbol(InputFile *I, std::string SymbolName,
```
`std::string` -> `const std::string &` or `llvm::StringRef`.

Also, should we rename the function to `addUndefinedSymbol`? We only support ELF target, and we typically do not use `ELF` for the other `NamePool` APIs.

quic-seaswara replied:
Sure

quic-seaswara replied:
Addressed the comments, thank  you!

partaror reviewed (commented):
Overall looks good. Some initial comments.

partaror commented:
We are getting an incorrect multiple definition error with this change.

Reproducible example:

```bash
#!/usr/bin/env bash

cat >1.c <<\EOF
int foo() { return 1; }
int v = 11;
EOF

cat >script.t <<\EOF
SECTIONS {
  u = v;
  v = 13;
  w = v;
}
EOF

LDs=(ld.eld ld.lld arm-none-eabi-ld.bfd)
SFs=(eld lld bfd)

clang -o 1.o 1.c -c -ffunction-sections -fdata-sections

for i in "${!SFs[@]}"; do
  ${LDs[$i]} -o 1.${SFs[$i]}.out 1.o -T script.t -Map 1.${SFs[$i]}.map.txt
done
```

partaror reviewed (changes requested):
The changes in this pull-request breaks compatibility with the previous behavior:

```bash
#!/usr/bin/env bash

cat >1.c <<\EOF
int foo = 11;
EOF

cat >script.t <<\EOF
PROVIDE(u = foo);
foo = 13 + u;
EOF

clang -o 1.o 1.c -c -ffunction-sections
ld.eld -o a.eld.out 1.o -T script.t -Map a.eld.map.txt
```

Previously for this link, eld used to (correctly) report the error:

```
Fatal: script.t: Error: undefined symbol 'u' referenced in expression
Fatal: Linking had errors.
```

eld with this patch does not report any error for this link.

quic-seaswara commented on include/eld/Diagnostics/DiagSymbolResolutions.inc:
```diff
@@ -94,3 +94,5 @@ DIAG(error_patchable_script, DiagnosticEngine::Error,
 DIAG(error_patchable_alias_not_found, DiagnosticEngine::Error,
      "Patchable alias '%0' was not found or was garbage collected")
+DIAG(warning_override_symbol, DiagnosticEngine::Warning,
+     "Symbol %0 defined in %1 overrides symbol in %2")
```
Sure

partaror commented:
> This is one of the behaviors that is needed by picolibc. This patch fixes this to reflect bfd behavior.

Oh, this is interesting. LLD errors out with the below message for this case:

```
ld.lld:  error: assignment to symbol foo does not converge
```

GNU ld links fine.

partaror commented:
This is another interesting (_weird_) case:

```bash
#!/usr/bin/env bash

cat >1.c <<\EOF
int foo = 11;
int v = 1;
EOF

cat >script.t <<\EOF
PROVIDE(u = 13);
foo = u + v;
PROVIDE(u = 15);
EOF

clang -o 1.o 1.c -c -ffunction-sections
LDs=(ld.eld ld.lld arm-none-eabi-ld.bfd)
SFs=(eld lld bfd)

for i in "${!SFs[@]}"; do
  ${LDs[$i]} -o a.${SFs[$i]}.out 1.o -T script.t -Map a.${SFs[$i]}.map.txt
done
```

Both lld and GNU ld selects `PROVIDE(u = 15);` for obtaining the value of `u`, however, eld selects the first `PROVIDE` statement: `PROVIDE(u = 13);`. If we are following GNU semantics, then should we select the second `PROVIDE` statement as well in this case?

quic-seaswara commented:
> This is another interesting (_weird_) case:
> 
> ```shell
> #!/usr/bin/env bash
> 
> cat >1.c <<\EOF
> int foo = 11;
> int v = 1;
> EOF
> 
> cat >script.t <<\EOF
> PROVIDE(u = 13);
> foo = u + v;
> PROVIDE(u = 15);
> EOF
> 
> clang -o 1.o 1.c -c -ffunction-sections
> LDs=(ld.eld ld.lld arm-none-eabi-ld.bfd)
> SFs=(eld lld bfd)
> 
> for i in "${!SFs[@]}"; do
>   ${LDs[$i]} -o a.${SFs[$i]}.out 1.o -T script.t -Map a.${SFs[$i]}.map.txt
> done
> ```
> 
> Both lld and GNU ld selects `PROVIDE(u = 15);` for obtaining the value of `u`, however, eld selects the first `PROVIDE` statement: `PROVIDE(u = 13);`. If we are following GNU semantics, then should we select the second `PROVIDE` statement as well in this case?

This is an old behavior, and requires a seperate issue. I believe we issue a warning, we should continue to do that as an improvement over GNU so that user knows and fix the error.

partaror commented:
> I believe we issue a warning, we should continue to do that as an improvement over GNU so that user knows and fix the error.

We do not report any warning :').

quic-seaswara commented:
> ```shell
> !/usr/bin/env bash
> 
> cat >1.c <<\EOF
> int foo = 11;
> int v = 1;
> EOF
> 
> cat >script.t <<\EOF
> PROVIDE(u = 13);
> foo = u + v;
> PROVIDE(u = 15);
> EOF
> 
> clang -o 1.o 1.c -c -ffunction-sections
> LDs=(ld.eld ld.lld arm-none-eabi-ld.bfd)
> SFs=(eld lld bfd)
> 
> for i in "${!SFs[@]}"; do
>   ${LDs[$i]} -o a.${SFs[$i]}.out 1.o -T script.t -Map a.${SFs[$i]}.map.txt
> done
> ```

```
$ ld.eld 1.o -T script.t  -Wall
Warning: Symbol foo defined in script.t overrides symbol in 1.o
Warning: script.t: Ignoring provide symbol 'u' redeclaration
```

quic-seaswara commented on lib/Script/Assignment.cpp:
```diff
@@ -264,3 +256,19 @@ bool Assignment::isDot() const { return (Name.size() == 1 && Name[0] == '.'); }
 }
+
+// Add undefined symbols for symbols referred by the assignment
+void Assignment::processAssignment(Module &CurModule, InputFile &I) {
```
It needs to be `processAssignment` as the function just does not add the symbols referenced by expression.

quic-seaswara commented on lib/Script/Assignment.cpp:
```diff
@@ -264,3 +256,19 @@ bool Assignment::isDot() const { return (Name.size() == 1 && Name[0] == '.'); }
+  if (isProvideOrProvideHidden())
+    return;
+  for (auto &Sym : getSymbolNames()) {
+    CurModule.getNamePool().addUndefinedELFSymbol(&I, Sym);
```
This is very interesting. The end behavior is the same. This requires more changes than the code in the patch, and I feel that this needs to be handled seperately. 

Could you please open an issue ?

quic-seaswara commented:
I think I have addressed most of all comments.

There is definitely room for improvement in the documentation aspect. 

I will take this as the next step.
